### DISCUSSION TITLE
Official solution for custom links (regexp)

### DISCUSSION THREAD

**Discussion by afska** [sourcelair/xterm.js#5223] [Q&A] [answered]
Hi. Thanks for this great project. I wanted to ask if we could get an official solution to register links with any format.

Like https://github.com/LabhanshAgrawal/xterm-link-provider but... official 😅 The problem with that project is that it's not updated anymore, it uses an old Xtermjs version and it has some [issues](https://github.com/LabhanshAgrawal/xterm-link-provider/pull/35) when handling line breaks & emojis (which I attempt to solve in that PR) but also incompatibilities with `addon-fit` (like links not formatting well if they end in the last character of a line).

I ask primarily because maybe you have a more generic solution working and it'd be a matter of just exposing it. I think a lot of users will find it useful!

**Comment by jerch** [comment, marked as answer]
Isn't that basically the same as the [web-links addon](https://github.com/xtermjs/xterm.js/tree/master/addons/addon-web-links)? Would the addon's API do for you?

Edit: Regarding your index issues in your PR - the index buffer mapping got reworked in the addon, as the old code (also partially used by xterm-link-provider) contained wrong assumptions about char progression on the terminal buffer.

**Comment by Tyriar** [comment]
Specifically this part of the API that lets you give your own regex:

https://github.com/xtermjs/xterm.js/blob/41e8ae395937011d6bf6c7cb618b851791aed395/addons/addon-web-links/typings/addon-web-links.d.ts#L48-L51

**Comment by afska** [comment]
Oh, cool, thanks for the help! I've used `WebLinkProvider` directly and it worked, and fixed the issues I was having with link parsing. I picked the provider and not the addon because I need to add different link providers and be able to dispose them.

The only issue I had was that this provider [assumes](https://github.com/xtermjs/xterm.js/blob/41e8ae395937011d6bf6c7cb618b851791aed395/addons/addon-web-links/src/WebLinkProvider.ts#L72) the regex has to be an URL. It makes sense, I guess that's why it's called _web_ link provider, heh. But:

In my use case, users are able to create custom links like file paths or even words/phrases that point to a dictionary. Removing that `if` made it possible. Since it was an easy solution, maybe that could be configurable? It would allow more flexible usages of the library like mine.

**Reply by Tyriar** [reply]
I'm not opposed to making that callback configurable, maybe by passing a custom callback to do the validation instead?

**Reply by afska** [reply]
Yeah, I think that'd be great!

**Reply by afska** [reply]
Can addons be added/disabled at any time? If not, would it be a good idea to also export WebLinkProvider in the library to allow this? (so users can call the dispose() method)

**Reply by rob-myers** [reply]
Thanks for pointing out that condition @afska . 
For the moment it's easy to patch the package to get things working.
